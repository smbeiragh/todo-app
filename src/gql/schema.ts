import { gql } from 'apollo-server-express';
import {FakeDB} from "./mockDB";

export const typeDefs = gql`
  type ToDo {
    id: String
    title: String
    priority: Int
    dueDate: String
    done: Boolean
  }
  
  type Query {
    getToDos(orderBy: String, orderType: String): [ToDo]
  }
  
  type Mutation {
    createToDo(title: String!, dueDate: String, priority: Int, orderBy: String, orderType: String): [ToDo]
    deleteToDo(id: String!): String
    updateToDo(id: String!, title: String, dueDate: String, priority: Int, done: Boolean): ToDo
  }
`;

export const createResolvers = (fakeDB: FakeDB) => ({
  Query: {
    getToDos: (root: any, {orderBy = 'title', orderType = 'ASC'} = {}) => {
      return fakeDB.getToDos({orderBy, orderType});
    }
  },
  Mutation: {
    createToDo: (
        root: any,
        { title = '', priority = 1, dueDate, done = false ,orderBy = 'title', orderType= 'ASC'}
        : { title: string, priority: number, dueDate: string, done: boolean, orderBy: string, orderType: string}
      ) => {
      fakeDB.createToDo({
        title,
        priority,
        dueDate,
        done
      });
      return fakeDB.getToDos({orderBy, orderType});
    },
    deleteToDo: (root: any, { id = '' } = {}) => {
      return fakeDB.deleteToDo({id});
    },
    updateToDo: (root: any, { id ,title, priority, dueDate, done} : ITodo = ({} as any)) : ITodo => {
      return fakeDB.updateToDo({ id ,title, priority, dueDate, done});
    },
  }
});
