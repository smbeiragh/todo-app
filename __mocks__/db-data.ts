export function creatData(): Array<ITodoInput> {
  const toDos: Array<ITodoInput> = [
    {
      title: 'Todo 1',
      priority: 2,
      dueDate: '2018-09-27T10:53:48.248Z',
      done: false
    },
    {
      title: 'Todo 2',
      priority: 1,
      dueDate: '2018-09-27T10:53:48.248Z',
      done: false
    }
  ];
  return toDos;
}
