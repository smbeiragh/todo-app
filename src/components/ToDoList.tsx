import * as React from 'react';
import { ToDo } from './ToDo';

type Props = {
  toDos: ITodo[]
  refetch: () => void
  orderBy: string
};

export function ToDoList({ toDos, refetch, orderBy }: Props) {
 return (
   <ul className={'c-toDoList'}>
     { toDos.map((todo: any) => (
       <ToDo className={'c-toDoList__item'} key={todo.id} toDo={todo} refetch={refetch} orderBy={orderBy} />
     ))}
   </ul>
 );
}
