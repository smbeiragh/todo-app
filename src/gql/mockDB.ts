const defaultDate = () :string => (new Date((new Date()).setDate((new Date()).getDate()+7))).toISOString();

class ToDo {
  id: string;
  title: string;
  priority: number;
  dueDate: string;
  dueDateTimeStamp: number;
  done: boolean;

  constructor({ title = '', priority = 1, dueDate = defaultDate() , done = false }: ITodoInput) {
    this.id = require('crypto').randomBytes(10).toString('hex'),
      this.title = title,
      this.priority = priority,
      this.dueDate = dueDate,
      this.dueDateTimeStamp = +(new Date(dueDate));
    this.done = done
  }

  setDueDate(dueDate: string) {
    this.dueDate = dueDate;
    this.dueDateTimeStamp = +(new Date(dueDate));
    return this;
  }
}

function sort(fakeDB: Array<ToDo> ,{orderBy = 'title', orderType = 'ASC'} = {}) {
  const fieldName = orderBy === 'dueDate' ? 'dueDateTimeStamp' : orderBy;
  if (orderType === 'ASC') {
    return fakeDB.sort((a, b) => (a as any)[fieldName] < (b as any)[fieldName] ? -1 : 1 )
  }
  return fakeDB.sort((a, b) => (a as any)[fieldName] > (b as any)[fieldName] ? -1 : 1)
}

function defaultInitData() {
  const data: Array<ITodoInput> = [
    {
      title: 'Create a Readme',
      priority: 2,
      dueDate: (new Date()).toISOString(),
      done: false
    },
    {
      title: 'Finish The project',
      priority: 1,
      dueDate: (new Date()).toISOString(),
      done: false
    },
    {
      title: 'Host the project on github',
      priority: 1,
      dueDate: (new Date()).toISOString(),
      done: false
    }
  ];

  return data;
}

export class FakeDB {

  data: Array<ToDo>;

  constructor(initData: Array<ITodoInput> = defaultInitData()) {
    this.data = initData.map((item) :ToDo => { return new ToDo(item) });
  }

  getToDos({orderBy = 'title', orderType = 'ASC'} = {}) {
    return sort(this.data, {orderBy, orderType});
  }

  createToDo({ title = '', priority = 1, dueDate, done = false }: ITodoInput) {
    const todo = new ToDo({
      title,
      priority,
      dueDate,
      done
    });
    this.data.push(todo);
    return todo;
  }

  deleteToDo({ id = '' } = {}) {
    const index =  (this.data as any).findIndex((item: any)=> id === item.id);
    if(index > -1) {
      this.data.splice(index, 1);
      return id;
    }
    return null;
  }

  updateToDo({ id ,title, priority, dueDate, done}: ITodo) : ITodo {
    if(id) {
      const todo =  (this.data as any).find((item: any)=> id === item.id);
      if (!todo) {
        return null;
      }
      if (title) {
        todo.title = title;
      }
      if (typeof priority === 'number') {
        todo.priority = priority;
      }
      if (dueDate) {
        todo.setDueDate(dueDate);
      }
      if (typeof done === "boolean") {
        todo.done = done;
      }

      return todo;
    }
    return null;
  }
}