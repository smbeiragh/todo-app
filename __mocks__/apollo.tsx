import * as path from 'path';
import { Server } from 'http';
import express from 'express';
import { createResolvers, typeDefs } from "../src/gql/schema";
import { ApolloServer } from "apollo-server-express";
import { ApolloProvider, renderToStringWithData } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { createHttpLink } from 'apollo-link-http';
import { InMemoryCache } from "apollo-cache-inmemory";
import * as React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import fetch from 'node-fetch';
import { rootReducer } from '../src/redux/reducers/index';
import { testAction } from '../src/redux/actions/index';
import { FakeDB } from '../src/gql/mockDB';

export function createServer(fakeDB: FakeDB) {
  const app = express();
  const server = new Server(app);

  const gqlServer = new ApolloServer({
    // These will be defined for both new or existing servers
    typeDefs,
    resolvers: createResolvers(fakeDB),
  });

  gqlServer.applyMiddleware({ app });

  // start the server
  const port = process.env.NODE_PORT;

  return new Promise((resolve: any, reject: any) => {
    server.listen(port, (err: any) => {
      if (err) {
        console.error(err);
        reject(err);
      } else {
        resolve(server);
        console.log(process.env.NODE_ENV);
        console.info(`Server is running on http://localhost:${port}`);
      }
    });
  });

}

export function createClient() {
  const store = createStore(rootReducer);

  // start the server
  const port = process.env.NODE_PORT;

  // lets put something in redux store
  store.dispatch(testAction({text: 'some test string!'}));

  const initState = JSON.stringify(store.getState());

  const apolloClient = new ApolloClient({
    ssrMode: true,
    // Remember that this is the interface the SSR server will use to connect to the
    // API server, so we need to ensure it isn't firewalled, etc
    link: createHttpLink({
      uri: `http://127.0.0.1:${port}/graphql`,
      // credentials: 'same-origin',
      fetch : fetch as any
    }),
    cache: new InMemoryCache(),
  });

  return { store, apolloClient };
}


