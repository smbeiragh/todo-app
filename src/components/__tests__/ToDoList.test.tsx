import * as React from 'react';
import { shallow } from 'enzyme';
import { ToDoList } from '../ToDoList';
import {FakeDB} from './../../gql/mockDB';
import { creatData } from '../../../__mocks__/db-data';

test('ToDo should render', () => {
  const db = new FakeDB(creatData());
  const toDos = db.getToDos();
  const component = shallow(
    <ToDoList toDos={toDos} refetch={()=>{}} orderBy={'title'}/>
  );
  expect(component).toMatchSnapshot();
});
