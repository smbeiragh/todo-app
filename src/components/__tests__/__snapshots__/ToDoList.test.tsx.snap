// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ToDoList
    orderBy="title"
    refetch={[Function]}
    toDos={
      Array [
        ToDo {
          "done": false,
          "dueDate": "2018-09-27T10:53:48.248Z",
          "dueDateTimeStamp": 1538045628248,
          "id": "d8a01e139ee9e81bd897",
          "priority": 2,
          "title": "Todo 1",
        },
        ToDo {
          "done": false,
          "dueDate": "2018-09-27T10:53:48.248Z",
          "dueDateTimeStamp": 1538045628248,
          "id": "040798423bed67f20f85",
          "priority": 1,
          "title": "Todo 2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ToDo
          className="c-toDoList__item"
          orderBy="title"
          refetch={[Function]}
          toDo={
            ToDo {
              "done": false,
              "dueDate": "2018-09-27T10:53:48.248Z",
              "dueDateTimeStamp": 1538045628248,
              "id": "d8a01e139ee9e81bd897",
              "priority": 2,
              "title": "Todo 1",
            }
          }
        />,
        <ToDo
          className="c-toDoList__item"
          orderBy="title"
          refetch={[Function]}
          toDo={
            ToDo {
              "done": false,
              "dueDate": "2018-09-27T10:53:48.248Z",
              "dueDateTimeStamp": 1538045628248,
              "id": "040798423bed67f20f85",
              "priority": 1,
              "title": "Todo 2",
            }
          }
        />,
      ],
      "className": "c-toDoList",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "d8a01e139ee9e81bd897",
        "nodeType": "function",
        "props": Object {
          "className": "c-toDoList__item",
          "orderBy": "title",
          "refetch": [Function],
          "toDo": ToDo {
            "done": false,
            "dueDate": "2018-09-27T10:53:48.248Z",
            "dueDateTimeStamp": 1538045628248,
            "id": "d8a01e139ee9e81bd897",
            "priority": 2,
            "title": "Todo 1",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "040798423bed67f20f85",
        "nodeType": "function",
        "props": Object {
          "className": "c-toDoList__item",
          "orderBy": "title",
          "refetch": [Function],
          "toDo": ToDo {
            "done": false,
            "dueDate": "2018-09-27T10:53:48.248Z",
            "dueDateTimeStamp": 1538045628248,
            "id": "040798423bed67f20f85",
            "priority": 1,
            "title": "Todo 2",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ToDo
            className="c-toDoList__item"
            orderBy="title"
            refetch={[Function]}
            toDo={
              ToDo {
                "done": false,
                "dueDate": "2018-09-27T10:53:48.248Z",
                "dueDateTimeStamp": 1538045628248,
                "id": "d8a01e139ee9e81bd897",
                "priority": 2,
                "title": "Todo 1",
              }
            }
          />,
          <ToDo
            className="c-toDoList__item"
            orderBy="title"
            refetch={[Function]}
            toDo={
              ToDo {
                "done": false,
                "dueDate": "2018-09-27T10:53:48.248Z",
                "dueDateTimeStamp": 1538045628248,
                "id": "040798423bed67f20f85",
                "priority": 1,
                "title": "Todo 2",
              }
            }
          />,
        ],
        "className": "c-toDoList",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "d8a01e139ee9e81bd897",
          "nodeType": "function",
          "props": Object {
            "className": "c-toDoList__item",
            "orderBy": "title",
            "refetch": [Function],
            "toDo": ToDo {
              "done": false,
              "dueDate": "2018-09-27T10:53:48.248Z",
              "dueDateTimeStamp": 1538045628248,
              "id": "d8a01e139ee9e81bd897",
              "priority": 2,
              "title": "Todo 1",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "040798423bed67f20f85",
          "nodeType": "function",
          "props": Object {
            "className": "c-toDoList__item",
            "orderBy": "title",
            "refetch": [Function],
            "toDo": ToDo {
              "done": false,
              "dueDate": "2018-09-27T10:53:48.248Z",
              "dueDateTimeStamp": 1538045628248,
              "id": "040798423bed67f20f85",
              "priority": 1,
              "title": "Todo 2",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
